pkg/metrics/producers/scheduledcapacity/crontabs.go

```diff
@@ -0,0 +1,77 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduledcapacity
+
+import (
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/autoscaling/v1alpha1"
+	"github.com/robfig/cron/v3"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Crontab string
+
+func (tab *Crontab) String() string {
+	return string(*tab)
+}
+
+// crontabFrom returns a crontab equivalent to the strongly-typed schedule
+func crontabFrom(pattern *v1alpha1.Pattern) *Crontab {
+	c := Crontab(fmt.Sprintf("%s %s %s %s %s", strings.Join(crontabFieldFrom(pattern.Minutes, true), ","),
+		strings.Join(crontabFieldFrom(pattern.Hours, true), ","), strings.Join(crontabFieldFrom(pattern.Days, false), ","),
+		strings.Join(crontabFieldFrom(pattern.Months, false), ","), strings.Join(crontabFieldFrom(pattern.Weekdays, false), ",")))
+	return &c
+}
+
+// crontabFieldFrom returns a field of a strongly-typed format into a crontab field
+func crontabFieldFrom(field *string, isZeroDefault bool) []string {
+	if field != nil {
+		elements := strings.Split(*field, ",")
+		for i, val := range elements {
+			// if the string input is not an integer
+			if _, err := strconv.Atoi(val); err != nil {
+				elements[i] = strings.ToUpper(val)
+			} else {
+				elements[i] = val
+			}
+			elements[i] = strings.Trim(val, " ")
+		}
+		return elements
+	} else {
+		// If minutes/hours field
+		if isZeroDefault {
+			return []string{"0"}
+		} else {
+			return []string{"*"}
+		}
+	}
+}
+
+// nextTime returns the next time that crontab will match in timezone loc
+func (tab *Crontab) nextTime(loc *time.Location) (time.Time, error) {
+	c := cron.New(cron.WithLocation(loc))
+	_, err := c.AddFunc(tab.String(), func() { fmt.Printf("crontab %s has been initialized", tab.String()) })
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not parse crontab: %w", err)
+	}
+
+	c.Start()
+	nextTime := c.Entries()[0].Next
```

Comment:
Can `Entries()` return an slice with len==0, in that case program will crash here?

Reply:
`Entries()` is directly correlated with the functions you add into crontab `c`. Since we add a function right above, It's safe to assume that it won't have length 0. If it doesn't properly add a function, the error will be handled.